File: go/tabnas_plugin_test.go
// Copyright (c) 2024-2026 Richard Rodger, MIT License

package tabnasjsonic_test

// The relaxed-JSON grammar is shipped as an idiomatic tabnas plugin:
// tabnas.Make().Use(jsonic.Grammar). These external tests pin that
// contract — the thing other grammar plugins depend on — independently of
// the legacy jsonic.Make() API.

import (
	"reflect"
	"strings"
	"testing"

	jsonic "github.com/tabnas/jsonic/go"
	tabnas "github.com/tabnas/parser/go"
)

func TestGrammarPluginOnBareEngine(t *testing.T) {
	j := tabnas.Make()
	if err := j.Use(jsonic.Grammar); err != nil {
		t.Fatal(err)
	}

	out, err := j.Parse("a:1,b:[x,y,z],c:{d:e} // tail")
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"a": float64(1),
		"b": []any{"x", "y", "z"},
		"c": map[string]any{"d": "e"},
	}
	if !reflect.DeepEqual(out, want) {
		t.Errorf("relaxed parse: got %#v, want %#v", out, want)
	}

	// Path diving and plain JSON also work.
	if out, _ := j.Parse("a:b:c:1"); !reflect.DeepEqual(out,
		map[string]any{"a": map[string]any{"b": map[string]any{"c": float64(1)}}}) {
		t.Errorf("path dive: got %#v", out)
	}
}

func TestGrammarPluginAppliesBranding(t *testing.T) {
	// The engine ships the relaxed lexer defaults; the plugin layers on the
	// jsonic error identity, so failures carry the [jsonic/...] tag.
	j := tabnas.Make()
	_ = j.Use(jsonic.Grammar)
	_, err := j.Parse(`"unterminated`)
	if err == nil {
		t.Fatal("expected error")
	}
	if !strings.Contains(err.Error(), "[jsonic/unterminated_string]") {
		t.Errorf("expected jsonic-branded error, got:\n%s", err.Error())
	}
}

func TestGrammarPluginLayeredDependency(t *testing.T) {
	// A second plugin builds on the grammar jsonic registered — the
	// "play nice as a dependency" use case. It adds keyword values on top
	// of jsonic's map/value rules. Register jsonic first.
	yesno := func(j *tabnas.Tabnas, _ map[string]any) error {
		j.SetOptions(tabnas.Options{
			Value: &tabnas.ValueOptions{
				Def: map[string]*tabnas.ValueDef{
					"yes": {Val: true},
					"no":  {Val: false},
				},
			},
		})
		return nil
	}

	j := tabnas.Make()
	_ = j.Use(jsonic.Grammar)
	_ = j.Use(yesno)

	out, err := j.Parse("a:yes,b:no,c:[yes,no]")
	if err != nil {
		t.Fatal(err)
	}
	want := map[string]any{
		"a": true,
		"b": false,
		"c": []any{true, false},
	}
	if !reflect.DeepEqual(out, want) {
		t.Errorf("layered parse: got %#v, want %#v", out, want)
	}
}

File: ts/test/tabnas-plugin.test.js
/* Copyright (c) 2024-2026 Richard Rodger and other contributors, MIT License */
'use strict'

// The relaxed-JSON grammar is shipped as an idiomatic `tabnas` plugin:
// `new Tabnas().use(jsonic)`. These tests pin that contract — the thing
// other grammar plugins depend on — independently of the legacy callable
// `Jsonic` API.

const { describe, it } = require('node:test')
const assert = require('node:assert')

const {
  Tabnas,
  jsonic,
  registerJsonicGrammar,
} = require('..')

describe('tabnas-plugin', function () {
  it('use-on-bare-engine', () => {
    const p = new Tabnas().use(jsonic)

    // Relaxed JSON: unquoted keys, implicit map, lists, comments, dive.
    assert.deepEqual(p.parse('a:1,b:[x,y,z],c:{d:e} // tail'), {
      a: 1,
      b: ['x', 'y', 'z'],
      c: { d: 'e' },
    })
    assert.deepEqual(p.parse('a:b:c:1'), { a: { b: { c: 1 } } })

    // Plain JSON still works.
    assert.deepEqual(p.parse('{"a":1,"b":[2,3]}'), { a: 1, b: [2, 3] })
  })

  it('applies-jsonic-branding', () => {
    // The engine ships the relaxed lexer defaults; the plugin layers on
    // jsonic's own error identity and hints.
    const p = new Tabnas().use(jsonic)
    assert.equal(p.options.errmsg.name, 'jsonic')
    assert.equal(p.options.errmsg.link, 'https://github.com/tabnas/jsonic')
  })

  it('plugins-at-construction', () => {
    const p = new Tabnas({ plugins: [jsonic] })
    assert.deepEqual(p.parse('x:1 y:2'), { x: 1, y: 2 })
  })

  it('layered-dependency', () => {
    // A second plugin builds on the grammar jsonic registered — the
    // "play nice as a dependency" use case. It adds a keyword value on
    // top of jsonic's map/value rules. Register jsonic first.
    function yesno(tn) {
      tn.options({ value: { def: { yes: { val: true }, no: { val: false } } } })
    }
    const p = new Tabnas().use(jsonic).use(yesno)
    assert.deepEqual(p.parse('a:yes,b:no,c:[yes,no]'), {
      a: true,
      b: false,
      c: [true, false],
    })
  })

  it('register-grammar-only', () => {
    // The grammar-only helper installs rules without re-applying options,
    // for plugins that manage their own option set.
    const p = new Tabnas()
    registerJsonicGrammar(p)
    assert.deepEqual(p.parse('a:1,b:2'), { a: 1, b: 2 })
  })

  it('plugin-name-is-jsonic', () => {
    // The plugin's function name drives its options namespace key.
    assert.equal(jsonic.name, 'jsonic')
  })
})
